cli/flox/src/commands/init/node.rs

```diff
@@ -285,20 +297,66 @@ impl Node {
     }
 
     /// Try to find nodejs (alone) that satisfies constraints in package.json
+    #[instrument(skip_all, fields(progress = "searching for compatible node versions"))]
     async fn try_find_compatible_nodejs(
         flox: &Flox,
         versions: &PackageJSONVersionsUnresolved,
     ) -> Result<Option<PackageJSONVersion>> {
         let PackageJSONVersionsUnresolved { node, .. } = versions;
 
+        debug!(
+            version = node.as_ref().unwrap_or(&"unspecified".to_string()),
+            "trying to find compatible node version"
+        );
+        // Prioritize newer major versions by putting them in order. The bare "nodejs"
+        // packages goes last because an old nodejs version that matches against it will
+        // pin the package group to an old page.
+        let nodejs_packages = ["nodejs_23", "nodejs_22", "nodejs_20", "nodejs_18", "nodejs"];
         let found_node = match node {
             Some(node_version) => {
-                match try_find_compatible_package(flox, "nodejs", Some(node_version)).await? {
-                    None => Some(PackageJSONVersion::Unavailable),
-                    Some(result) => Some(PackageJSONVersion::Found(result)),
+                let resolution_futs = nodejs_packages.iter().map(|attr_path| {
+                    try_find_compatible_package(flox, attr_path, Some(node_version))
+                });
+                // The result of `join_all` is a `Vec` of the result of resolving each
+                // future in the order that they were provided e.g. the order of
+                // `nodejs_packages`. This means the results are returned in order of
+                // preference for latest package version *when using the live catalog*.
+                // However, since the order in which the futures are resolved at runtime
+                // isn't deterministic, the order in which the resolutions are recorded
+                // in the test JSON file is also not deterministic. That means you may get
+                // results returned in the wrong order during tests. To mitigate that
+                // we just put the results in a map and pull them back out in the
+                // correct order.
+                let resolutions = join_all(resolution_futs)
+                    .await
+                    .into_iter()
+                    .filter_map(|res| res.unwrap_or_default())
+                    .map(|pkg| (pkg.attr_path.to_string(), pkg))
+                    .collect::<BTreeMap<String, _>>();
+                let resolutions_sorted = {
+                    let mut buf = Vec::new();
+                    for pkg in nodejs_packages.iter() {
+                        if let Some(result) = resolutions.get(*pkg) {
```

Comment:
question nonblocking: I'm curious how come you decided to do everything serverside rather than pre-filtering `nodejs_packages` clientside?

Reply:
I went this way because our canonical version parsing is handled on the server, and didn't want to reimplement it for something non-trivial like `<=X, >Y` in order to know which request to send. The catalog's primary purpose is doing this exact thing -> converting name/version constraints into a thing a user can install, so I figured I would leverage that.